apps/mobile/src/features/send/components/recipient/v2/build-recipient-suggestions.ts

```diff
@@ -0,0 +1,179 @@
+  getAddressByAccount: (fingerprint: string, accountIndex: number) => string;
+  performBnsLookup: (name: string) => Promise<string | null>;
+  validateAddress: (value: string) => Promise<boolean>;
+}
+
+export async function buildRecipientSuggestions({
+  searchTerm,
+  accounts,
+  selectedAccount,
+  activity,
+  canSelfSend,
+  findAccountByAddress,
+  getAddressByAccount,
+  performBnsLookup,
+  validateAddress,
+}: BuildRecipientSuggestionsParams): Promise<RecipientSection[]> {
+  const normalizedSearchTerm = searchTerm.trim().toLowerCase();
+  const recentEntries = getRecents({ activity, findAccountByAddress });
+  const accountEntries = getAccounts({
+    accounts,
+    selectedAccount,
+    canSelfSend,
+    getAddressByAccount,
+  });
+
+  function getSearchResults() {
+    return performSearch({
+      searchTerm: normalizedSearchTerm,
+      staticEntries: [...recentEntries, ...accountEntries],
+      performBnsLookup,
+      validateAddress,
+    });
+  }
+
+  const sections =
+    normalizedSearchTerm.length > 0
+      ? [createSection('matching', await getSearchResults())]
+      : [createSection('recents', recentEntries), createSection('accounts', accountEntries)];
+  return sections.filter(section => section.data.length > 0);
+}
+
+interface PerformSearchParams {
+  searchTerm: string;
+  staticEntries: RecipientSuggestionEntry[];
+  performBnsLookup: (name: string) => Promise<string | null>;
+  validateAddress: (value: string) => Promise<boolean>;
+}
+
+async function performSearch({
+  searchTerm,
+  staticEntries,
+  performBnsLookup,
+  validateAddress,
+}: PerformSearchParams): Promise<RecipientSuggestionEntry[]> {
+  if (searchTerm.length === 0) {
+    return [];
+  }
+
+  if (await validateAddress(searchTerm)) {
+    const existingEntry = staticEntries.find(isEntryWithSameAddress(searchTerm));
+    return existingEntry ? [existingEntry] : [createExternalEntry(searchTerm)];
+  }
+
+  const matches = staticEntries.filter(isLooseNameOrAddressMatch(searchTerm));
+  if (matches.length > 0) {
+    return matches;
+  }
+
+  if (isBnsLookupCandidate(searchTerm)) {
+    const resolvedAddress = await performBnsLookup(searchTerm);
+    return resolvedAddress ? [createExternalEntry(resolvedAddress, searchTerm)] : [];
+  }
+
+  return [];
+}
+
+interface GetRecentsParams {
+  activity: SendAssetActivity[];
+  findAccountByAddress: (address: string) => Account | null;
+}
+
+export function getRecents({
+  activity,
+  findAccountByAddress,
+}: GetRecentsParams): RecipientSuggestionEntry[] {
+  const recentItemLimit = 4;
+
+  return pipe(
+    activity,
+    filter(activity => activity.receivers.length > 0),
+    // Filter out contract targets
+    filter(activity => activity.receivers.some(r => !r.includes('.'))),
+    // Deliberately filter out own accounts from Recents. Unclear how to handle taproot addresses matching existing accounts.
+    filter(activity => findAccountByAddress(activity.receivers[0] ?? '') == null),
```

Comment:
Missing `!`. Worth enabling https://eslint.org/docs/latest/rules/eqeqeq, didn't realize it wasn't in the default preset.